### ISSUE TITLE
How to freeze backbone when finetune on custom dataset

### ISSUE THREAD

**Issue by lzyuan168** [tinyvision/DAMO-YOLO#126]
### Before Asking

- [X] I have read the [README](https://github.com/tinyvision/DAMO-YOLO/blob/master/README.md) carefully. 我已经仔细阅读了README上的操作指引。

- [X] I want to train my custom dataset, and I have read the [tutorials for finetune on your data](https://github.com/tinyvision/DAMO-YOLO/blob/master/README.md#Finetune%20on%20your%20data) carefully and organize my dataset correctly; 我想训练自定义数据集，我已经仔细阅读了训练自定义数据的教程，以及按照正确的目录结构存放数据集。

- [X] I have pulled the latest code of main branch to run again and the problem still existed. 我已经拉取了主分支上最新的代码，重新运行之后，问题仍不能解决。

### Search before asking

- [X] I have searched the DAMO-YOLO [issues](https://github.com/tinyvision/DAMO-YOLO/issues) and found no similar questions.

### Question

Hi, I would like to know if there is an option to freeze backbone when finetuning on custom dataset. Because according to the the tutorial, there is only a line to add "self.train.finetune_path" which points to the pretrained .pth weights. However, if I run the train.py file without freezing the backbone, isn't that equal to training from scratch?

Thank you.

**Comment by ksv87** [comment]
https://github.com/tinyvision/DAMO-YOLO/blob/319572eef367340267ab6ab8ae253527a71d7c3a/damo/detectors/detector.py#L12-L25

you can freeze backbone here

**Comment by nuhash-bcraft** [comment]
> [DAMO-YOLO/damo/detectors/detector.py](https://github.com/tinyvision/DAMO-YOLO/blob/319572eef367340267ab6ab8ae253527a71d7c3a/damo/detectors/detector.py#L12-L25)
> 
> Lines 12 to 25 in [319572e](/tinyvision/DAMO-YOLO/commit/319572eef367340267ab6ab8ae253527a71d7c3a)
> 
>   
>   
>  class Detector(nn.Module): 
>      def __init__(self, config): 
>          super().__init__() 
>   
>          self.backbone = build_backbone(config.model.backbone) 
>          self.neck = build_neck(config.model.neck) 
>          self.head = build_head(config.model.head) 
>   
>          self.config = config 
>   
>      def init_bn(self, M): 
>   
> you can freeze backbone here



Also had to update build optimizer.

def build_optimizer(self, models, cfg, exp_module=None):
        if not isinstance(models, (tuple, list)):
            models = (models, )

        param_dict = {}
        base_wd = cfg.get('weight_decay', None)
        optimizer_name = cfg.pop('name')
        optim_cls = getattr(torch.optim, optimizer_name)
        for model in models:
            for name, param in model.named_parameters():
                if not param.requires_grad:
                    continue
                assert param not in param_dict
                param_dict[param] = {"name": name}

            ##weight decay of bn is always 0.
            # for name, m in model.named_modules():
            #     if isinstance(m, NORMS):
            #         if hasattr(m, "bias") and m.bias is not None:
            #             param_dict[m.bias].update({"weight_decay": 0})
            #         param_dict[m.weight].update({"weight_decay": 0})
            
            # # Ensure weight decay for BatchNorm layers is set to 0 safely
            for name, m in model.named_modules():
                if isinstance(m, NORMS):  # Assuming NORMS = (nn.BatchNorm2d, nn.BatchNorm1d, etc.)
                    if hasattr(m, "bias") and m.bias is not None:
                        if m.bias in param_dict:  # Check if bias exists in param_dict
                            param_dict[m.bias].update({"weight_decay": 0})
                    if m.weight in param_dict:  # Check if weight exists in param_dict
                        param_dict[m.weight].update({"weight_decay": 0})


            # weight decay of bias is always 0.
            # for name, m in model.named_modules():
            #     if hasattr(m, "bias") and m.bias is not None:
            #         param_dict[m.bias].update({"weight_decay": 0})
            
            for name, m in model.named_modules():
                if hasattr(m, "bias") and m.bias is not None:
                    if m.bias in param_dict:  # Ensure bias exists before updating
                        param_dict[m.bias].update({"weight_decay": 0})
            param_groups = []
        for p, pconfig in param_dict.items():
            name = pconfig.pop("name", None)
            param_groups += [{"params": p, **pconfig}]


        optimizer = optim_cls(param_groups, **cfg)
        print("optimizer:",optimizer)

        return optimizer

**Comment by nuhash-bcraft** [comment]
Update the code for optimizer too.

def build_optimizer(self, models, cfg, exp_module=None):
        if not isinstance(models, (tuple, list)):
            models = (models, )

        param_dict = {}
        base_wd = cfg.get('weight_decay', None)
        optimizer_name = cfg.pop('name')
        optim_cls = getattr(torch.optim, optimizer_name)
        for model in models:
            for name, param in model.named_parameters():
                if not param.requires_grad:
                    continue
                assert param not in param_dict
                param_dict[param] = {"name": name}

            ##weight decay of bn is always 0.
            # for name, m in model.named_modules():
            #     if isinstance(m, NORMS):
            #         if hasattr(m, "bias") and m.bias is not None:
            #             param_dict[m.bias].update({"weight_decay": 0})
            #         param_dict[m.weight].update({"weight_decay": 0})
            
            # # Ensure weight decay for BatchNorm layers is set to 0 safely
            for name, m in model.named_modules():
                if isinstance(m, NORMS):  # Assuming NORMS = (nn.BatchNorm2d, nn.BatchNorm1d, etc.)
                    if hasattr(m, "bias") and m.bias is not None:
                        if m.bias in param_dict:  # Check if bias exists in param_dict
                            param_dict[m.bias].update({"weight_decay": 0})
                    if m.weight in param_dict:  # Check if weight exists in param_dict
                        param_dict[m.weight].update({"weight_decay": 0})


            # weight decay of bias is always 0.
            # for name, m in model.named_modules():
            #     if hasattr(m, "bias") and m.bias is not None:
            #         param_dict[m.bias].update({"weight_decay": 0})
            
            for name, m in model.named_modules():
                if hasattr(m, "bias") and m.bias is not None:
                    if m.bias in param_dict:  # Ensure bias exists before updating
                        param_dict[m.bias].update({"weight_decay": 0})
            param_groups = []
        for p, pconfig in param_dict.items():
            name = pconfig.pop("name", None)
            param_groups += [{"params": p, **pconfig}]


        optimizer = optim_cls(param_groups, **cfg)
        print("optimizer:",optimizer)

        return optimizer
